doc: Move from Make to CMake

carlescufi opened the pull request:
Tested on Linux and Windows 10.

Fixes #8284

carlescufi committed:
doc: Support building with CMake

Add a new CMakeLists.txt that is capable of building the documentation.
Note that the error checkin currently only works on Unix platforms since
it's tied to a shell script.

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

carlescufi commented:
@dbkinder @nashif I am missing the translation of that `filter-doc-log.sh` script.
Also I'd like to know if the rest of the targets are worthwhile, or which are.

carlescufi commented:
CC @ru-fu

dbkinder commented:
The important targets are "htmldocs"  (and its internal dependencies "html", "content", "kconfig", and "doxy" ) and "clean".  The other output targets for latex, json, htmlhelp, singlehtml, applehelp, and such aren't used (they come from the autogenerated Makefile created when Sphinx is initially set up.

This change will also impact the top-level Makefile, documentation for how to generate the Zephyr docs, and the CI processes that are using "make htmldocs" to generate the nightly doc build.

carlescufi commented:
@dbkinder I’ll fix that so you can run the same Make command from the root

carlescufi commented:
@dbkinder I will keep a couple of the targets that seem to have been used since they’ve had their dependencies added. I will drop the rest.

tejlmand commented on doc/CMakeLists.txt:
```diff
@@ -0,0 +1,89 @@
+
+set(DOXYFILE ${CMAKE_BINARY_DIR}/build.doxyfile)
+
+set_property(SOURCE APPEND PROPERTY CMAKE_CONFIGURE_DEPENDS ${DOXYFILE})
```
This comment covers the lines: 36-38:
You can use configure_file instead.
Advantage of configure_file is that it will setup the CMake dependency for you as well.

> `set_property(SOURCE APPEND PROPERTY CMAKE_CONFIGURE_DEPENDS ${DOXYFILE})`
> `file(READ ${CMAKE_CURRENT_SOURCE_DIR}/zephyr.doxyfile CONTENTS)`
> `file(WRITE ${DOXYFILE} ${CONTENTS})`

can be written as:
`configure_file(${CMAKE_CURRENT_SOURCE_DIR}/zephyr.doxyfile ${DOXYFILE} COPYONLY)`

and then you just append to the file, as you do now with:
> `file(APPEND ${DOXYFILE} "STRIP_FROM_PATH=${ZEPHYR_BASE}")`

carlescufi committed:
doc: Change Makefile and doc for building docs to CMake

Now that CMake is supported for building the docs, adapt the
instructions to build the documentation to the new mechanism and also
adapt the root-level Makefile to support building using CMake.

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

carlescufi committed:
doc: Remove old Makefile

With the transition to CMake, the old Makefile is not required anymore.

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

carlescufi commented:
@tejlmand comments addressed
@dbkinder added support for building using the Makefile and also updated the doc

carlescufi commented on doc/CMakeLists.txt:
```diff
@@ -0,0 +1,105 @@
+  COMMAND ${CMAKE_COMMAND} -E remove ${CMAKE_CURRENT_SOURCE_DIR}/${DOC_LOG}
+)
+
+#add_dependencies(clean pristine)
```
done

tejlmand reviewed (approved):
Looking good

carlescufi committed:
cmake: util: Add process.cmake script

In order to be able to execute a process at build time (and not at
configure time) in a cross-platform manner while at the same time
redirecting stdout and stderr to files, add a script that wraps CMake's
execute_process() command.

The script can be invoked with identical parameters to execute_process()
but at build time, allowing CMake users to store stdout and stderr in
files in a way that works on all operating systems.

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

carlescufi committed:
cmake: util: Add fmerge.cmake script

In order to merge multiple files into a single one, this convenience
script is provided to be able to do so at build time in a manner
compatible with all operating systems.

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

carlescufi committed:
scripts: Print results from filter-known-issues.py

In order to get rid of doc/scripts/filter-doc-log.sh to make building
the documentation cross-platform, include the functionality in there in
the already existing filter-known-issues.py.

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

carlescufi committed:
doc: Remove unused filter-doc-log.sh

The filter-doc-log.sh script is no longer used since its functionality
has been moved to the CMakeLists.txt and filter-known-issues.py.

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

dbkinder reviewed (approved):
seems to work OK :)
It may complicate things when we upgrade to a new version of Sphinx, but we'll deal with that.

dbkinder commented on doc/README.rst:
```diff
@@ -82,18 +82,43 @@ Running the documentation processors
+.. code-block:: console
+
+   cd $ZEPHYR_BASE
+   make htmldocs
```
still need to run

    source zephyr-env.sh

before building the docs.

carlescufi replied:
@dbkinder fixed, thanks!

dbkinder reviewed (changes requested):
missed one comment

carlescufi commented:
>seems to work OK :)

More importantly, works on Windows :)

>It may complicate things when we upgrade to a new version of Sphinx, but we'll deal with that.

I hope not, the invocation of Sphinx is now actually cleaner than before, I can definitely help you out with the `CMakeLists.txt` modifications when we upgrade if you need help.

dbkinder reviewed (approved):
Thanks.
